### ISSUE TITLE
Deluge exporter 100% CPU

### ISSUE THREAD

**Issue by ngosang** [tobbez/deluge_exporter#23]
Deluge exporter: 2.4.0

I stopped Deluge container for maintenance during 10 minutes and Deluge exporter CPU went 100% for several hours.
I had to restart Deluge exporter to make it work again. In the traces I can see it can't connect to Deluge, that's right, but we should handle those situations and add an sleep for a few seconds or something like that to avoid wasting resources. The most problematic for me was it wasn't able to recover the connection after Deluge came back.

![image](https://user-images.githubusercontent.com/10577978/190279269-0b6bd467-0141-4c05-95ff-2e733bcbfad1.png)

```
  | 2022-08-28 23:16:29 | {"log":"2022-08-28 23:16:29.602 \| ERROR    \| prometheus_client.registry:collect:83 - An error has been caught in function 'collect', process 'MainProcess' (1), thread 'Thread-230 (process_request_thread)' (548336180000):\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605463667Z"}
  | 2022-08-28 23:16:29 | {"log":"Traceback (most recent call last):\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605624925Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/threading.py\", line 966, in _bootstrap\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605659962Z"}
  | 2022-08-28 23:16:29 | {"log":"    self._bootstrap_inner()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605682443Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/threading.py\", line 1009, in _bootstrap_inner\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605706813Z"}
  | 2022-08-28 23:16:29 | {"log":"    self.run()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605730739Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/threading.py\", line 946, in run\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.60575272Z"}
  | 2022-08-28 23:16:29 | {"log":"    self._target(*self._args, **self._kwargs)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605774331Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/socketserver.py\", line 683, in process_request_thread\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605795034Z"}
  | 2022-08-28 23:16:29 | {"log":"    self.finish_request(request, client_address)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605886237Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/socketserver.py\", line 360, in finish_request\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605915959Z"}
  | 2022-08-28 23:16:29 | {"log":"    self.RequestHandlerClass(request, client_address, self)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605937626Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/socketserver.py\", line 747, in __init__\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605958163Z"}
  | 2022-08-28 23:16:29 | {"log":"    self.handle()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.605978866Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/wsgiref/simple_server.py\", line 134, in handle\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606000625Z"}
  | 2022-08-28 23:16:29 | {"log":"    handler.run(self.server.get_app())\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.60602131Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/wsgiref/handlers.py\", line 137, in run\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606042347Z"}
  | 2022-08-28 23:16:29 | {"log":"    self.result = application(self.environ, self.start_response)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606062995Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/site-packages/prometheus_client/exposition.py\", line 118, in prometheus_app\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.60608455Z"}
  | 2022-08-28 23:16:29 | {"log":"    status, header, output = _bake_output(registry, accept_header, params)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606106161Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/site-packages/prometheus_client/exposition.py\", line 100, in _bake_output\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606126828Z"}
  | 2022-08-28 23:16:29 | {"log":"    output = encoder(registry)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606148235Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/site-packages/prometheus_client/openmetrics/exposition.py\", line 21, in generate_latest\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606169457Z"}
  | 2022-08-28 23:16:29 | {"log":"    for metric in registry.collect():\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606191049Z"}
  | 2022-08-28 23:16:29 | {"log":"\u003e File \"/usr/local/lib/python3.10/site-packages/prometheus_client/registry.py\", line 83, in collect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606214993Z"}
  | 2022-08-28 23:16:29 | {"log":"    yield from collector.collect()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606236808Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/deluge_exporter.py\", line 88, in collect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606258012Z"}
  | 2022-08-28 23:16:29 | {"log":"    client.connect()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606278956Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/site-packages/deluge_client/client.py\", line 84, in connect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606299122Z"}
  | 2022-08-28 23:16:29 | {"log":"    self._connect()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606321641Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/site-packages/deluge_client/client.py\", line 98, in _connect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606347122Z"}
  | 2022-08-28 23:16:29 | {"log":"    self._socket.connect((self.host, self.port))\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606371733Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/ssl.py\", line 1374, in connect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606393233Z"}
  | 2022-08-28 23:16:29 | {"log":"    self._real_connect(addr, False)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606465213Z"}
  | 2022-08-28 23:16:29 | {"log":"  File \"/usr/local/lib/python3.10/ssl.py\", line 1361, in _real_connect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.60649075Z"}
  | 2022-08-28 23:16:29 | {"log":"    super().connect(addr)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606512231Z"}
  | 2022-08-28 23:16:29 | {"log":"socket.gaierror: [Errno -2] Name does not resolve\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:29.606531916Z"}
  | 2022-08-28 23:16:59 | {"log":"2022-08-28 23:16:59.600 \| ERROR    \| prometheus_client.registry:collect:83 - An error has been caught in function 'collect', process 'MainProcess' (1), thread 'Thread-231 (process_request_thread)' (548336180000):\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603091676Z"}
  | 2022-08-28 23:16:59 | {"log":"Traceback (most recent call last):\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603238693Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/threading.py\", line 966, in _bootstrap\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603266878Z"}
  | 2022-08-28 23:16:59 | {"log":"    self._bootstrap_inner()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603289026Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/threading.py\", line 1009, in _bootstrap_inner\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.60330997Z"}
  | 2022-08-28 23:16:59 | {"log":"    self.run()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603330748Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/threading.py\", line 946, in run\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603350636Z"}
  | 2022-08-28 23:16:59 | {"log":"    self._target(*self._args, **self._kwargs)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603371155Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/socketserver.py\", line 683, in process_request_thread\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603390988Z"}
  | 2022-08-28 23:16:59 | {"log":"    self.finish_request(request, client_address)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603411284Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/socketserver.py\", line 360, in finish_request\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603430802Z"}
  | 2022-08-28 23:16:59 | {"log":"    self.RequestHandlerClass(request, client_address, self)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603450598Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/socketserver.py\", line 747, in __init__\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603470357Z"}
  | 2022-08-28 23:16:59 | {"log":"    self.handle()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603490005Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/wsgiref/simple_server.py\", line 134, in handle\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603509838Z"}
  | 2022-08-28 23:16:59 | {"log":"    handler.run(self.server.get_app())\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603530061Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/wsgiref/handlers.py\", line 137, in run\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603550023Z"}
  | 2022-08-28 23:16:59 | {"log":"    self.result = application(self.environ, self.start_response)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603570486Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/site-packages/prometheus_client/exposition.py\", line 118, in prometheus_app\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603590541Z"}
  | 2022-08-28 23:16:59 | {"log":"    status, header, output = _bake_output(registry, accept_header, params)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603612745Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/site-packages/prometheus_client/exposition.py\", line 100, in _bake_output\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603684263Z"}
  | 2022-08-28 23:16:59 | {"log":"    output = encoder(registry)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603713244Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/site-packages/prometheus_client/openmetrics/exposition.py\", line 21, in generate_latest\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603734559Z"}
  | 2022-08-28 23:16:59 | {"log":"    for metric in registry.collect():\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603756781Z"}
  | 2022-08-28 23:16:59 | {"log":"\u003e File \"/usr/local/lib/python3.10/site-packages/prometheus_client/registry.py\", line 83, in collect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603776799Z"}
  | 2022-08-28 23:16:59 | {"log":"    yield from collector.collect()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603798373Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/deluge_exporter.py\", line 88, in collect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603818021Z"}
  | 2022-08-28 23:16:59 | {"log":"    client.connect()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603838039Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/site-packages/deluge_client/client.py\", line 84, in connect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603857928Z"}
  | 2022-08-28 23:16:59 | {"log":"    self._connect()\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603878354Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/site-packages/deluge_client/client.py\", line 98, in _connect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603902761Z"}
  | 2022-08-28 23:16:59 | {"log":"    self._socket.connect((self.host, self.port))\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603924316Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/ssl.py\", line 1374, in connect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603944446Z"}
  | 2022-08-28 23:16:59 | {"log":"    self._real_connect(addr, False)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603964686Z"}
  | 2022-08-28 23:16:59 | {"log":"  File \"/usr/local/lib/python3.10/ssl.py\", line 1361, in _real_connect\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.603984556Z"}
  | 2022-08-28 23:16:59 | {"log":"    super().connect(addr)\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.604004834Z"}
  | 2022-08-28 23:16:59 | {"log":"socket.gaierror: [Errno -2] Name does not resolve\n","stream":"stderr","attrs":{"tag":"deluge-exporter"},"time":"2022-08-28T21:16:59.604024463Z"}
```

**Comment by tobbez** [comment]
There is no reconnection logic in deluge_exporter. It makes a single connection attempt once on startup, and once each time it's scraped.

The problem you report seems to be an issue in the deluge-client library, and there's a commit in an open PR that should fix it: https://github.com/JohnDoee/deluge-client/pull/42/commits/6dee9eae981f3dabef9155b02fd12345d0a8ee40.

Closing this for now, as there's not much I can do. When/if deluge-client releases a version containing a fix, the deluge_exporter containers should automatically get it within a week.

Additionally, the logs you provided are unrelated (to the CPU usage issue). Regardless, I've improved the behavior for that case (a single line is now logged, not a full traceback: https://github.com/tobbez/deluge_exporter/commit/09cf3c0923ba723d74c69c6a903129905310f7d1).

If those logs are from after deluge was started again, then the failure to connect is a different issue, and one that's outside the control of the exporter. The error message indicates a problem with the environment where the exporter is running (docker?), causing lookups of the deluge hostname to fail.
